// @elgato-stream-deck/core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEVICE_MODELS = exports.DeviceModelType = exports.VENDOR_ID = exports.StreamDeckProxy = void 0;
const id_1 = require("./id");
const models_1 = require("./models");
const plus_1 = require("./models/plus");
__exportStar(require("./types"), exports);
__exportStar(require("./id"), exports);
var proxy_1 = require("./proxy");
Object.defineProperty(exports, "StreamDeckProxy", { enumerable: true, get: function () { return proxy_1.StreamDeckProxy; } });
/** Elgato vendor id */
exports.VENDOR_ID = 0x0fd9;
var DeviceModelType;
(function (DeviceModelType) {
    DeviceModelType["STREAMDECK"] = "streamdeck";
    DeviceModelType["PEDAL"] = "pedal";
})(DeviceModelType = exports.DeviceModelType || (exports.DeviceModelType = {}));
/** List of all the known models, and the classes to use them */
exports.DEVICE_MODELS = [
    {
        id: id_1.DeviceModelId.ORIGINAL,
        type: DeviceModelType.STREAMDECK,
        productId: 0x0060,
        class: models_1.StreamDeckOriginal,
    },
    {
        id: id_1.DeviceModelId.MINI,
        type: DeviceModelType.STREAMDECK,
        productId: 0x0063,
        class: models_1.StreamDeckMini,
    },
    {
        id: id_1.DeviceModelId.XL,
        type: DeviceModelType.STREAMDECK,
        productId: 0x006c,
        class: models_1.StreamDeckXL,
    },
    {
        id: id_1.DeviceModelId.ORIGINALV2,
        type: DeviceModelType.STREAMDECK,
        productId: 0x006d,
        class: models_1.StreamDeckOriginalV2,
    },
    {
        id: id_1.DeviceModelId.ORIGINALMK2,
        type: DeviceModelType.STREAMDECK,
        productId: 0x0080,
        class: models_1.StreamDeckOriginalMK2,
    },
    {
        id: id_1.DeviceModelId.PLUS,
        type: DeviceModelType.STREAMDECK,
        productId: 0x0084,
        class: plus_1.StreamDeckPlus,
    },
    {
        id: id_1.DeviceModelId.PEDAL,
        type: DeviceModelType.PEDAL,
        productId: 0x0086,
        class: models_1.StreamDeckPedal,
    },
    {
        id: id_1.DeviceModelId.XLV2,
        type: DeviceModelType.STREAMDECK,
        productId: 0x008f,
        class: models_1.StreamDeckXLV2,
    },
    {
        id: id_1.DeviceModelId.MINIV2,
        type: DeviceModelType.STREAMDECK,
        productId: 0x0090,
        class: models_1.StreamDeckMiniV2,
    },
];
